### ISSUE TITLE
some details

### ISSUE THREAD

**Issue by laomao0** [sniklaus/softmax-splatting#12]
Hi, Snikalus,
     I have 2 questions about the details of your network.

<img width="727" alt="Screen Shot 2020-05-26 at 10 22 33 AM" src="https://user-images.githubusercontent.com/40321136/82854370-d2191480-9f3a-11ea-8c2c-66958e8ab39a.png">

   1. (a) In this picture, the warped frames and warped first-level features are concatenated [torch.cat], right?  (b) If it is right? for example, for the first reference image I_0, and first-level features of I_0, they have (32+3) channels, right?  (c) If (b) is right, the input to the gridnet is concatenate{warped_I0, warped_feature_I0, wapred_I1, wapred_feature_I1 }, totally 70 channels, right? and the initial LateralBlock of gridnet down the 70 channel to 32 channel. Is my understanding, right?

2. the second question is about feature warping. your network extracts three pyramid features.  For example, if input img is 1x3x256x256 [N C H W], the first-level feature is [1x32x256x256], the second-level feature is [1x64x128x128], the third-level feature is [1x96x64x64]. When warping the second-level and thrid level feature, we need the optical flow F. The initial flow is [1x2x256x256], you down-sample the flow F to [1x2x128x128] to warp the second-level flow, right? for example? nn.functional.interpolate(metric0, scale_factor=0.5, mode = 'bilinear', align_corners = False)

<img width="392" alt="Screen Shot 2020-05-26 at 10 16 53 AM" src="https://user-images.githubusercontent.com/40321136/82854106-07713280-9f3a-11ea-9734-742d30ce2554.png">

thanks for your reply~

**Comment by sniklaus** [comment]
Yes, the warped frames and the warped first-level features are concatenated. Yes, the input to the green lateral block on the top left is 70 channels (3 + 3 + 32 + 32). Yes, the output of the block on the top left is 32 (the number in each block denotes the number of output channels).

Yes, the optical flow needs to be downsampled. Note that it is common for optical flow estimators based on deep learning to yield a low-res prediction (in the case for PWC-Net, the estimate is 1/4th of the input resolution that is then upsampled to get the full-resolution prediction). So you may not have to downsample the flow for the coarse levels but upsample the flow for the fine levels.

**Comment by laomao0** [comment]
thanks for your reply!

**Comment by laomao0** [comment]
> Yes, the warped frames and the warped first-level features are concatenated. Yes, the input to the green lateral block on the top left is 70 channels (3 + 3 + 32 + 32). Yes, the output of the block on the top left is 32 (the number in each block denotes the number of output channels).
> 
> Yes, the optical flow needs to be downsampled. Note that it is common for optical flow estimators based on deep learning to yield a low-res prediction (in the case for PWC-Net, the estimate is 1/4th of the input resolution that is then upsampled to get the full-resolution prediction). So you may not have to downsample the flow for the coarse levels but upsample the flow for the fine levels.

thanks for your reply!
Using PWC-net, when I upsample the flow of the 1/4 the resolution to 1/1 resolution (for example 10x10 -> 40x40 pixels), the multiplication factor is 20 in the PWC-net code [line 305 https://github.com/sniklaus/pytorch-pwc/blob/master/run.py].  

If I need to up-sample the 1/4 flow to 1/2 resolution (for example flow of 10x10 -> 20 x 20), the factor is 20/2=10. Is it right? 

Thanks for your patience.

**Comment by laomao0** [comment]
> Yes, the warped frames and the warped first-level features are concatenated. Yes, the input to the green lateral block on the top left is 70 channels (3 + 3 + 32 + 32). Yes, the output of the block on the top left is 32 (the number in each block denotes the number of output channels).
> 
> Yes, the optical flow needs to be downsampled. Note that it is common for optical flow estimators based on deep learning to yield a low-res prediction (in the case for PWC-Net, the estimate is 1/4th of the input resolution that is then upsampled to get the full-resolution prediction). So you may not have to downsample the flow for the coarse levels but upsample the flow for the fine levels.

-------------------------
thanks for your reply!
Using PWC-net, when I upsample the flow of the 1/4 the resolution to 1/1 resolution (for example 10x10 -> 40x40 pixels), the multiplication factor is 20 in the PWC-net code [line 305 https://github.com/sniklaus/pytorch-pwc/blob/master/run.py].

If I need to up-sample the 1/4 flow to 1/2 resolution (for example flow of 10x10 -> 20 x 20), the factor is 20/2=10. Is it right?

Thanks for your patience.

**Comment by sniklaus** [comment]
You are correct indeed. I would not necessarily do the interpolation the way you outlined though. The predicted `meta_flow` may already be at the resolution of `tensorFlow_L3`, so interpolation may not be necessary. As for `tensorFlow_L2`, I would compute it from `2 * upsample(tensorFlow_L3)` and likewise for `tensorFlow_L1`, I would compute it from `2 * upsample(tensorFlow_L2)`.

**Comment by laomao0** [comment]
> You are correct indeed. I would not necessarily do the interpolation the way you outlined though. The predicted `meta_flow` may already be at the resolution of `tensorFlow_L3`, so interpolation may not be necessary. As for `tensorFlow_L2`, I would compute it from `2 * upsample(tensorFlow_L3)` and likewise for `tensorFlow_L1`, I would compute it from `2 * upsample(tensorFlow_L2)`.

Thanks for your reply, that helps me a lot.
